monero-sys/src/lib.rs

```diff
@@ -0,0 +1,1511 @@
+    pub async fn unlocked_balance(&self) -> monero::Amount {
+        self.call(move |wallet| wallet.unlocked_balance()).await
+    }
+
+    /// Get the total balance of the wallet.
+    pub async fn total_balance(&self) -> monero::Amount {
+        self.call(move |wallet| wallet.total_balance()).await
+    }
+
+    /// Check if the wallet is synchronized.
+    async fn synchronized(&self) -> bool {
+        self.call(move |wallet| wallet.synchronized()).await
+    }
+
+    /// Get the sync progress of the wallet.
+    async fn sync_progress(&self) -> SyncProgress {
+        self.call(move |wallet| wallet.sync_progress()).await
+    }
+
+    /// Check if the wallet is connected to a daemon.
+    pub async fn connected(&self) -> bool {
+        self.call(move |wallet| wallet.connected()).await
+    }
+
+    /// Check that the wallet is created and ready to use.
+    /// Call this after creating a wallet to make sure the wallet thread responds correctly.
+    async fn check_wallet(&self) -> anyhow::Result<()> {
+        let (sender, receiver) = oneshot::channel();
+
+        self.call_sender
+            .send(Call {
+                function: Box::new(move |wallet| {
+                    Box::new(wallet.check_error()) as Box<dyn Any + Send>
+                }),
+                sender,
+            })
+            .map_err(|_| anyhow::anyhow!("failed to send check_wallet call"))?;
+
+        receiver
+            .await
+            .context("wallet channel closed unexpectedly")?;
+
+        Ok(())
+    }
+
+    /// Allow the wallet to connect to a daemon with a different version.
+    /// Also trusts the daemon.
+    /// Only used for regtests.
+    /// Also forces a full sync, which is only feasible in regtests.
+    #[doc(hidden)]
+    pub async fn unsafe_prepare_for_regtest(&self) {
+        self.call(move |wallet| {
+            wallet.force_full_sync();
+            wallet.allow_mismatched_daemon_version();
+            wallet.set_trusted_daemon(true);
+        })
+        .await
+    }
+
+    /// Wait until the wallet is synchronized.
+    ///
+    /// Polls the wallet's sync status every 500ms until the wallet is synchronized.
+    ///
+    /// If a listener is provided, it will be called with the sync progress.
+    pub async fn wait_until_synced(
+        &self,
+        listener: Option<impl Fn(SyncProgress) + Send + 'static>,
+    ) -> anyhow::Result<()> {
+        // We wait for ms before polling the wallet's sync status again.
+        // This is ok because this doesn't involve any blocking calls.
+        const POLL_INTERVAL_MILLIS: u64 = 500;
+
+        // Initiate the sync (make sure to drop the lock right after)
+        {
+            self.call(move |wallet| {
+                wallet.start_refresh();
+                wallet.refresh_async();
+            })
+            .await;
+            tracing::debug!("Wallet refresh initiated");
+        }
+
+        // Wait until the wallet is connected to the daemon.
+        loop {
```

Comment:
This will never terminate if we cannot connect to the daemon. Is this okay?

Either terminate here after a timeout **or** (even better) let this be handled by the C++ code. I'm assuming they will failed at some point if they cannot connect.